### PR TITLE
TP + FSDP distributed training (full finetuning)

### PR DESCRIPTION
**Pull request by acisseJZhong**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
Enabled TP training on top of FSDP. This is both from user request https://github.com/pytorch/torchtune/issues/2280, and also to improve memory so that we can get rid of `fsdp_cpu_offload` when training larger model. 

* **Reference PRs for this work: https://github.com/pytorch/torchtune/pull/2204, https://github.com/pytorch/torchtune/pull/2318**

Note: 
- TP + FSDP doesn't work with fused optimizer yet(`optimizer.fused` needs to be False). 
- Applying TP needs to happen after AC since AC changes the fqn. (this causes a few hours debugging :( )

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Running TP = 2, FSPD = 4:
image
image
Running TP = 1, FSDP = 8:
image
image
#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -502,6 +512,27 @@ def _setup_model(
         if self._compile:
             training.compile_model(model, verbose=self._is_rank_zero)
 
+        self.device_mesh = dist.init_device_mesh(
```
Comment (joecummings) [inline recipes/full_finetune_distributed.py:515]: I would rather explicitly save what we need here. So instead of putting the whole device_mesh on the object, later on if data_parallel_dim > 1, save self.local_size = dp_mesh.size() and self.local_rank = dp_mesh.get_local_rank() else self.local_size = self.world_size and self.local_rank = self.rank.

Does that make sense?

**Review by joecummings — APPROVED** [review]
Looks awesome! Just need to make sure the merge changes are correct.
